test/builtin_dataset_mocks.py

```diff
@@ -0,0 +1,492 @@
+from torchvision.prototype.datasets._api import find
+from torchvision.prototype.datasets.utils._internal import add_suggestion
+
+make_tensor = functools.partial(_make_tensor, device="cpu")
+
+__all__ = ["load"]
+
+DEFAULT_TEST_DECODER = object()
+
+
+class DatasetMocks:
+    def __init__(self):
+        self._mock_data_fns = {}
+        self._tmp_home = pathlib.Path(tempfile.mkdtemp())
+        self._cache = {}
+
+    def register_mock_data_fn(self, mock_data_fn):
+        name = mock_data_fn.__name__
+        if name not in datasets.list():
+            raise pytest.UsageError(
+                add_suggestion(
+                    f"The name of the mock data function '{name}' has no corresponding dataset.",
+                    word=name,
+                    possibilities=datasets.list(),
+                    close_match_hint=lambda close_match: f"Did you mean to name it '{close_match}'?",
+                    alternative_hint=lambda _: "",
+                )
+            )
+        self._mock_data_fns[name] = mock_data_fn
+        return mock_data_fn
+
+    def _parse_mock_info(self, mock_info, *, name):
+        if mock_info is None:
+            raise pytest.UsageError(
+                f"The mock data function for dataset '{name}' returned nothing. It needs to at least return an integer "
+                f"indicating the number of samples for the current `config`."
+            )
+        elif isinstance(mock_info, int):
+            mock_info = dict(num_samples=mock_info)
+        elif not isinstance(mock_info, dict):
+            raise pytest.UsageError(
+                f"The mock data function for dataset '{name}' returned a {type(mock_info)}. The returned object should "
+                f"be a dictionary containing at least the number of samples for the current `config` for the key "
+                f"`'num_samples'`. If no additional information is required for specific tests, the number of samples "
+                f"can also be returned as an integer."
+            )
+        elif "num_samples" not in mock_info:
+            raise pytest.UsageError(
+                f"The dictionary returned by the mock data function for dataset '{name}' must contain a `'num_samples'` "
+                f"entry indicating the number of samples for the current `config`."
+            )
+        return mock_info
+
+    def _get(self, dataset, config):
+        name = dataset.info.name
+        resources_and_mock_info = self._cache.get((name, config))
+        if resources_and_mock_info:
+            return resources_and_mock_info
+
+        try:
+            fakedata_fn = self._mock_data_fns[name]
+        except KeyError:
+            raise pytest.UsageError(
+                f"No mock data available for dataset '{name}'. "
+                f"Did you add a new dataset, but forget to provide mock data for it? "
+                f"Did you register the mock data function with `@DatasetMocks.register_mock_data_fn`?"
+            )
```

Comment:
> Perhaps I don't understand the decorators as well, but it looks like you put most of the mock data function here - won't this make this file huge and somewhat unreadable once we add more and more datasets?

Unfortunately, yes. 

> Should we perhaps add this mock data in the datasets themselves?

That is a fair question. In general I like the idea to have the mock data close the "actual" data. The problem I see, is that we would need to pull some pure test utilities such as `make_tar` or `make_image_folder` inside the `torchvision` package. I'm not opposed to that, but that would break the current paradigm that our tests are kind of "standalone".

PyTorch core also has a `testing` namespace with an `_internal` submodule where a lot of this functionality lives. We could also do that. In the light of #4721 we could also place some specialized comparison / generation functions there for downstream libraries to use. @NicolasHug